tools/schemacode/schemacode/tests/test_schema.py

```diff
@@ -27,3 +27,101 @@ def test_object_definitions(schema_obj):
+        "datetime": [
+            "2022-01-05T13:16:30",
+            "2022-01-05T13:16:30.5",  # subsecond resolution is allowed
+            "2022-01-05T13:16:30.000005",  # up to 6 decimal points
+            "2022-01-05T13:16:30UTC",  # timezones are allowed
+            "2022-01-05T13:16:30.05UTC",
+        ],
+        "time": [
+            "13:16:30",
+            "09:00:00",
+            "9:00:00",  # leading zeros are not required for hours
+        ],
+        "unit": ["any string is valid."],
+        "stimuli_relative": ["any/arbitrary/path/file.txt"],
+        "dataset_relative": ["any/arbitrary/path/file.txt"],
+        "participant_relative": ["any/arbitrary/path/file.txt"],
+        "rrid": ["RRID:SCR_017398"],
+        "uri": ["foo://example.com:8042/over/there?name=ferret#nose"],
+        "bids_uri": [
+            "bids::sub-01/fmap/sub-01_dir-AP_epi.nii.gz",
+            "bids:ds000001:sub-02/anat/sub-02_T1w.nii.gz",
+            "bids:myderivatives:sub-03/func/sub-03_task-rest_space-MNI152_bold.nii.gz",
+        ],
+    }
+    for pattern, test_list in GOOD_PATTERNS.items():
+        pattern_format = schema_obj["objects"]["formats"][pattern]["pattern"]
+        search_pattern = "^" + pattern_format + "$"
+        search = re.compile(search_pattern)
+        for test_string in test_list:
+            assert bool(
+                search.fullmatch(test_string)
+            ), f"'{test_string}' is not a valid match for the pattern '{search.pattern}'"
+
+    # Check that invalid strings do not match the search pattern.
+    BAD_PATTERNS = {
+        "label": ["test_01", "!", "010101-"],
+        "index": ["test", "0.1", "0-1", "0_1"],
+        "string": [],
+        "integer": ["3.14", "-3.14", "1.", "-1.", "string", "s1", "1%"],
+        "number": ["string", "1%"],
+        "boolean": ["True", "False", "T", "F"],
+        "date": [
+            "05-01-2022",  # MM-DD-YYYY or DD-MM-YYYY
+            "05/01/2022",  # MM/DD/YYYY or DD/MM/YYYY
+        ],
+        "datetime": [
+            "2022-01-05T13:16:30.1000005",  # too many decimal points
+            "2022-01-05T13:16:30U",  # time zone too short
+            "2022-01-05T13:16:30UTCUTC",  # time zone too long
+            "2022-01-05T34:10:10",  # invalid time
+        ],
+        "time": [
+            "34:10:10",  # invalid time
+        ],
+        "unit": [],
+        "stimuli_relative": [
+            "/path/with/starting/slash/file.txt",
+            "stimuli/path/file.txt",
+        ],
+        "dataset_relative": [
+            "/path/with/starting/slash/file.txt",
+        ],
+        "participant_relative": [
+            "/path/with/starting/slash/file.txt",
+            "sub-01/path/file.txt",
+        ],
+        "rrid": [],
+        "uri": [],
+        "bids_uri": [],
+    }
+    for pattern, test_list in BAD_PATTERNS.items():
+        pattern_format = schema_obj["objects"]["formats"][pattern]["pattern"]
+        search_pattern = "^" + pattern_format + "$"
```

Comment:
I guess I just love f-strings too much?

```suggestion
        search_pattern = f"^{pattern_format}$"
```